cpp/src/utilities/stream_pool.cpp

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <mutex>
+
+#include <cudf/detail/utilities/logger.hpp>
+#include <cudf/detail/utilities/stream_pool.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/error.hpp>
+
+namespace cudf::detail {
+
+namespace {
+
+// TODO: what is a good number here. what's the penalty for making it larger?
+// Dave Baranec rule of thumb was max_streams_needed * num_concurrent_threads,
+// where num_concurrent_threads was estimated to be 4. so using 32 will allow
+// for 8 streams per thread, which should be plenty (decoding will be up to 4
+// kernels when delta_byte_array decoding is added). rmm::cuda_stream_pool
+// defaults to 16.
+std::size_t constexpr STREAM_POOL_SIZE = 32;
+
+class rmm_cuda_stream_pool : public cuda_stream_pool {
+  rmm::cuda_stream_pool _pool;
+
+ public:
+  rmm_cuda_stream_pool() : _pool{STREAM_POOL_SIZE} {}
+  rmm::cuda_stream_view get_stream() override { return _pool.get_stream(); }
+  rmm::cuda_stream_view get_stream(std::size_t stream_id) override
+  {
+    return _pool.get_stream(stream_id);
+  }
+
+  std::vector<rmm::cuda_stream_view> get_streams(uint32_t count) override
+  {
+    static std::mutex stream_pool_mutex;
+
+    if (count > STREAM_POOL_SIZE) {
+      CUDF_LOG_WARN("get_streams called with count ({}) > pool size ({})", count, STREAM_POOL_SIZE);
+    }
+    auto streams = std::vector<rmm::cuda_stream_view>();
+    std::lock_guard<std::mutex> lock(stream_pool_mutex);
```

Comment:
I think I see a bit of a problem. This lock only protects against other `get_streams` calls, but concurrent `get_stream` calls can still "steal" a stream in the middle of the expected order.
I'm starting to think that we should drop the mutex for now. If this turns out to be important, maybe a case can be made to have the feature in the rmm pool.

Reply:
I still think that we only need a simple wrapper around the RMM stream pool that allows for debugging with 1 stream. I don't really understand why the locks are important. RMM already has an atomically incrementing counter for each call to `get_stream()` that ensures work is cycled across the stream pool regardless of the calling thread. https://github.com/rapidsai/rmm/blob/228330543cba4921e6b71a52dd1fa81c63921420/include/rmm/cuda_stream_pool.hpp#L96